File: air_quality_esp32/src/DeviceConfig.cpp

#include "DeviceConfig.h"
#include <ArduinoJson.h>

#define RECENT_DATA_MIN_PERIOD_SEC 30
#define RECENT_DATA_MAX_PERIOD_SEC 600

#define AQM_MEASUREMENTS_MIN_PERIOD_SEC 300
#define AQM_MEASUREMENTS_MAX_PERIOD_SEC 3600

#define JSON_HOSTNAME "hostname"
#define JSON_DEVICE_NAME "device_name"
#define JSON_RECENT_PERIOD "recent_data_period"
#define JSON_WIFI_NETWORKS "wifi_networks"
#define JSON_WIFI_SSID "ssid"
#define JSON_WIFI_PASSWORD "password"

#define JSON_AQM "aqm"
#define JSON_AQM_DEVICE_UUID "device_uuid"
#define JSON_AQM_BACKEND_ADDR "backend_addr"
#define JSON_AQM_API_KEY "api_key"
#define JSON_AQM_SAVE_MEASUREMENTS "save_measurements"
#define JSON_AQM_SEND_DATA "send_data"
#define JSON_AQM_MEASUREMENT_PERIOD "measurement_period"

#define CLAMP(x, min, max) (x > max ? max : (x < min ? min : x) )

bool DeviceConfig::parse(std::string configJson) {

  networks.clear();

  JsonDocument doc;

  auto error = deserializeJson(doc, configJson);
  if(error) {
    return false;
  }

  JsonObject aqm = doc[JSON_AQM];

  hostname = doc[JSON_HOSTNAME].as<std::string>();
  deviceName = doc[JSON_DEVICE_NAME].as<std::string>();

  if(hostname.empty() || deviceName.empty()) {
    return false;
  }

  uint32_t recentPeriod = doc[JSON_RECENT_PERIOD].as<uint32_t>();
  this->recentDataPeriod = CLAMP(recentPeriod, RECENT_DATA_MIN_PERIOD_SEC, RECENT_DATA_MAX_PERIOD_SEC);

  aqmConfig.deviceUuid = aqm[JSON_AQM_DEVICE_UUID].as<std::string>();
  aqmConfig.backendAddress = aqm[JSON_AQM_BACKEND_ADDR].as<std::string>();
  aqmConfig.apiKey = aqm[JSON_AQM_API_KEY].as<std::string>();

  aqmConfig.saveMeasurements = aqm[JSON_AQM_SAVE_MEASUREMENTS].as<bool>();
  aqmConfig.sendData = aqm[JSON_AQM_SEND_DATA].as<bool>();

  uint32_t aqmPeriod = aqm[JSON_AQM_MEASUREMENT_PERIOD].as<uint32_t>();
  aqmConfig.measurementPeriod = CLAMP(aqmPeriod, AQM_MEASUREMENTS_MIN_PERIOD_SEC, AQM_MEASUREMENTS_MAX_PERIOD_SEC);

  JsonArray networksJson = doc[JSON_WIFI_NETWORKS].as<JsonArray>();

  for(JsonObject e : networksJson) {
    WifiNetwork network;
    network.ssid = e[JSON_WIFI_SSID].as<std::string>();
    network.password = e[JSON_WIFI_PASSWORD].as<std::string>();

    if(network.ssid.empty()) {
      continue;
    }

    networks.push_back(network);
  }

  return true;
}

std::string DeviceConfig::toJson() {

  JsonDocument doc;

  doc[JSON_HOSTNAME] = hostname;
  doc[JSON_DEVICE_NAME] = deviceName;
  doc[JSON_RECENT_PERIOD] = recentDataPeriod;

  JsonObject aqm = doc[JSON_AQM].to<JsonObject>();
  aqm[JSON_AQM_DEVICE_UUID] = aqmConfig.deviceUuid;
  aqm[JSON_AQM_BACKEND_ADDR] = aqmConfig.backendAddress;
  aqm[JSON_AQM_API_KEY] = aqmConfig.apiKey;
  aqm[JSON_AQM_SAVE_MEASUREMENTS] = aqmConfig.saveMeasurements;
  aqm[JSON_AQM_SEND_DATA] = aqmConfig.sendData;
  aqm[JSON_AQM_MEASUREMENT_PERIOD] = aqmConfig.measurementPeriod;
  
  JsonArray networksJson = doc[JSON_WIFI_NETWORKS].to<JsonArray>();
  
  for(auto& net : networks) {
    JsonObject netJson = networksJson.add<JsonObject>();
    netJson[JSON_WIFI_SSID] = net.ssid;
    netJson[JSON_WIFI_PASSWORD] = net.password;
  }

  std::string configJson = "";
  serializeJson(doc, configJson);
  
  return configJson;
}

File: smart_home_app/air_quality/lib/src/models/device_config.dart
const _jsonHostname = "hostname";
const _jsonDeviceName = "device_name";
const _jsonRecentPeriod = "recent_data_period";
const _jsonWifiNetworks = "wifi_networks";
const _jsonWifiSsid = "ssid";
const _jsonWifiPassword = "password";

const _jsonAqm = "aqm";
const _jsonAqmDeviceUuid = "device_uuid";
const _jsonAqmBackendAddress = "backend_addr";
const _jsonAqmApiKey = "api_key";
const _jsonAqmSaveMeasurements = "save_measurements";
const _jsonAqmSendData = "send_data";
const _jsonAqmMeasurementPeriod = "measurement_period";

class WifiNetwork {
  String name;
  String password;

  WifiNetwork({required this.name, required this.password});

  factory WifiNetwork.fromJson(Map<String, dynamic> json) {
    return WifiNetwork(
      name: json[_jsonWifiSsid],
      password: json[_jsonWifiPassword],
    );
  }

  Map<String, dynamic> toJson() {
    return {_jsonWifiSsid: name, _jsonWifiPassword: password};
  }

  WifiNetwork clone() {
    return WifiNetwork(name: name, password: password);
  }
}

class AqmConfig {
  /// Device UUID
  String deviceUuid;

  /// Backend address
  String backendAddress;

  /// Api key
  String apiKey;

  /// Save air quality measurements or not
  bool saveMeasurements;

  /// Send measurements and logs to the backend or not
  bool sendData;

  /// Time period in seconds for saving measurements
  int measurementPeriod;

  AqmConfig({
    required this.deviceUuid,
    required this.backendAddress,
    required this.apiKey,
    required this.saveMeasurements,
    required this.sendData,
    required this.measurementPeriod,
  });

  factory AqmConfig.fromJson(Map<String, dynamic> json) {
    return AqmConfig(
      deviceUuid: json[_jsonAqmDeviceUuid],
      backendAddress: json[_jsonAqmBackendAddress],
      apiKey: json[_jsonAqmApiKey],
      saveMeasurements: json[_jsonAqmSaveMeasurements],
      sendData: json[_jsonAqmSendData],
      measurementPeriod: json[_jsonAqmMeasurementPeriod],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      _jsonAqmDeviceUuid: deviceUuid,
      _jsonAqmBackendAddress: backendAddress,
      _jsonAqmApiKey: apiKey,
      _jsonAqmSaveMeasurements: saveMeasurements,
      _jsonAqmSendData: sendData,
      _jsonAqmMeasurementPeriod: measurementPeriod,
    };
  }

  AqmConfig clone() {
    return AqmConfig(
      deviceUuid: deviceUuid,
      backendAddress: backendAddress,
      apiKey: apiKey,
      saveMeasurements: saveMeasurements,
      sendData: sendData,
      measurementPeriod: measurementPeriod,
    );
  }
}

class DeviceConfig {
  /// mDNS device hostname
  String hostname;

  /// Device name
  String deviceName;

  /// Time period in seconds for saving measurements to recent data
  int recentPeriod;

  AqmConfig aqmConfig;

  List<WifiNetwork> wifiNetworks;

  DeviceConfig({
    required this.hostname,
    required this.deviceName,
    required this.recentPeriod,
    required this.aqmConfig,
    required this.wifiNetworks,
  });

  void updateWifiNetworks(List<WifiNetwork> networks) {
    wifiNetworks = networks;
  }

  DeviceConfig clone() {
    return DeviceConfig(
      hostname: hostname,
      deviceName: deviceName,
      recentPeriod: recentPeriod,
      aqmConfig: aqmConfig.clone(),
      wifiNetworks: wifiNetworks.map((e) => e.clone()).toList(),
    );
  }

  factory DeviceConfig.fromJson(Map<String, dynamic> json) {
    return DeviceConfig(
      hostname: json[_jsonHostname],
      deviceName: json[_jsonDeviceName],
      recentPeriod: json[_jsonRecentPeriod],
      aqmConfig: AqmConfig.fromJson(json[_jsonAqm]),
      wifiNetworks: (json[_jsonWifiNetworks] as List<dynamic>)
          .map((e) => WifiNetwork.fromJson(e))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      _jsonHostname: hostname,
      _jsonDeviceName: deviceName,
      _jsonRecentPeriod: recentPeriod,
      _jsonAqm: aqmConfig.toJson(),
      _jsonWifiNetworks: wifiNetworks.map((e) => e.toJson()).toList(),
    };
  }
}
